### DISCUSSION TITLE
IPSec VPN failover in OpenWrt

### DISCUSSION THREAD

**Discussion by mrsagarjain1** [strongswan/strongswan#986] [Q&A]
I have two WAN connections in my OpenWrt router. WAN1 has metric 1 and WAN2 has metric 2. I have also configured mwan3 on my router and it is working fine. I have created an IPSec VPN from my OpenWrt router to the FortiGate firewall and it has been established and working fine. The problem is when WAN1 gets down the IPSec VPN doesn't shift to WAN2 because I think it follows the kernel routing table and tries to establish VPN through WAN1. When I manually down the WAN1 connection through` ifdown WAN1` command. The IPSec VPN gets established.

Below are the related files.

`cat /etc/ipsec.conf`
conn vpn1
    left=%any
    right=13.235.10.40
    leftsubnet=192.168.18.0/24
    leftfirewall=yes
    rightfirewall=no
    ikelifetime=24h
    lifetime=24h
    margintime=9m
    keyingtries=3
    dpdaction=none
    dpddelay=30s
    leftauth=psk
    rightauth=psk
    rightsubnet=172.16.1.0/24
    auto=start
    forceencaps=no
    keyexchange=ikev2
    esp=aes256-sha256-modp1536
    ike=aes256-sha256-modp1536
    type=tunnel

`cat /etc/config/mwan3`
config globals 'globals'
        option mmx_mask '0x3F00'
        option rtmon_interval '5'
        option mode 'mwan'

config member 'wan_mem'
        option interface 'wan'
        option metric '1'
        option weight '1'

config member 'sim_mem'
        option interface 'sim'
        option metric '2'
        option weight '1'

config interface 'wan'
        option enabled '1'
        option family 'ipv4'
        option track_method 'ping'
        option reliability '1'
        option max_ttl '60'
        option check_quality '0'
        option timeout '2'
        option interval '5'
        option failure_interval '5'
        option recovery_interval '5'
        option up '1'
        option initial_state 'offline'
        option size '8'
        option count '1'
        option down '3'
        list track_ip '8.8.8.8'
        list flush_conntrack 'ifup'
        list flush_conntrack 'ifdown'

config interface 'sim'
        option enabled '1'
        option family 'ipv4'
        option track_method 'ping'
        option reliability '1'
        option max_ttl '60'
        option check_quality '0'
        option timeout '2'
        option interval '5'
        option failure_interval '5'
        option recovery_interval '5'
        option up '1'
        option initial_state 'offline'
        option size '8'
        option count '1'
        option down '5'
        list track_ip '8.8.8.8'
        list flush_conntrack 'ifup'
        list flush_conntrack 'ifdown'

config policy 'wan_aggregation'
        list use_member 'wan_mem'
        list use_member 'sim_mem'
        option last_resort 'unreachable'

config rule 'default_rule'
        option sticky '0'
        option use_policy 'wan_aggregation'
        option proto 'all'
        
`cat /etc/config/network`
config interface 'wan'
        option ifname 'eth0.2'
        option proto 'dhcp'
        option metric '1'

config switch
        option name 'switch0'
        option reset '1'
        option enable_vlan '1'

config switch_vlan
        option device 'switch0'
        option vlan '1'
        option vid '1'
        option ports '6t 0 1 2'

config switch_vlan
        option device 'switch0'
        option vlan '2'
        option ports '6t 4'
        option vid '2'

config interface 'sim'
        option delegate '0'
        option proto '3g'
        option ipv6 'auto'
        option device '/dev/ttyUSB1'
        option service 'umts'
        option dialnumber '*99#'
        option sim1apn 'airtelgprs.com'
        option sim2apn 'airtelgprs.com'
        option apn 'airtelgprs.com'
        option defaultroute '0'

config switch_vlan
        option device 'switch0'
        option vlan '3'
        option ports '6t 3'
        option vid '3'

config route
        option target '0.0.0.0'
        option netmask '0.0.0.0'
        option interface 'sim'
        option metric '2'

`ipsec statusall`

Status of IKE charon daemon (strongSwan 5.8.2, Linux 4.14.241, mips):
  uptime: 5 minutes, since Apr 05 17:20:05 2022
  worker threads: 11 of 16 idle, 5/0/0/0 working, job queue: 0/0/0/0, scheduled: 3
  loaded plugins: charon aes des sha2 sha1 md4 md5 random nonce x509 pubkey gmp xcbc hmac kernel-netlink socket-default stroke updown
Listening IP addresses:
  192.168.18.1
  192.168.100.115
  10.128.114.102
Connections:
        vpn1:  %any...13.235.10.40  IKEv2
        vpn1:   local:  uses pre-shared key authentication
        vpn1:   remote: [13.235.10.40] uses pre-shared key authentication
        vpn1:   child:  192.168.18.0/24 === 172.16.1.0/24 TUNNEL
Security Associations (1 up, 0 connecting):
        vpn1[1]: ESTABLISHED 5 minutes ago, 10.128.114.102[10.128.114.102]...13.235.10.40[13.235.10.40]
        vpn1[1]: IKEv2 SPIs: 0bdf45855d0b1edb_i* caf70b1cb9b5448a_r, pre-shared key reauthentication in 23 hours
        vpn1[1]: IKE proposal: AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1536
        vpn1{1}:  INSTALLED, TUNNEL, reqid 1, ESP in UDP SPIs: c46d9bc7_i 52b2a38a_o
        vpn1{1}:  AES_CBC_256/HMAC_SHA2_256_128, 0 bytes_i, 11424 bytes_o (136 pkts, 191s ago), rekeying in 23 hours
        vpn1{1}:   192.168.18.0/24 === 172.16.1.0/24

`ip route show table all`

default via 192.168.100.1 dev eth0.2 table 1 metric 1
10.64.64.64 dev 3g-sim table 1 proto kernel scope link src 10.128.114.102
192.168.18.0/24 dev br-lan table 1 proto kernel scope link src 192.168.18.1
192.168.100.0/24 dev eth0.2 table 1 proto static scope link metric 1
default dev 3g-sim table 2 scope link
10.64.64.64 dev 3g-sim table 2 proto kernel scope link src 10.128.114.102
192.168.18.0/24 dev br-lan table 2 proto kernel scope link src 192.168.18.1
192.168.100.0/24 dev eth0.2 table 2 proto static scope link metric 1
172.16.1.0/24 via 13.235.10.40 dev 3g-sim table 220 proto static src 192.168.18.1
default via 192.168.100.1 dev eth0.2 proto static src 192.168.100.115 metric 1
default dev 3g-sim proto static scope link metric 2
10.64.64.64 dev 3g-sim proto kernel scope link src 10.128.114.102
192.168.18.0/24 dev br-lan proto kernel scope link src 192.168.18.1
192.168.100.0/24 dev eth0.2 proto static scope link metric 1
local 10.128.114.102 dev 3g-sim table local proto kernel scope host src 10.128.114.102
broadcast 127.0.0.0 dev lo table local proto kernel scope link src 127.0.0.1
local 127.0.0.0/8 dev lo table local proto kernel scope host src 127.0.0.1
local 127.0.0.1 dev lo table local proto kernel scope host src 127.0.0.1
broadcast 127.255.255.255 dev lo table local proto kernel scope link src 127.0.0.1
broadcast 192.168.18.0 dev br-lan table local proto kernel scope link src 192.168.18.1
local 192.168.18.1 dev br-lan table local proto kernel scope host src 192.168.18.1
broadcast 192.168.18.255 dev br-lan table local proto kernel scope link src 192.168.18.1
broadcast 192.168.100.0 dev eth0.2 table local proto kernel scope link src 192.168.100.115
local 192.168.100.115 dev eth0.2 table local proto kernel scope host src 192.168.100.115
broadcast 192.168.100.255 dev eth0.2 table local proto kernel scope link src 192.168.100.115
fe80::/64 dev eth0 proto kernel metric 256 pref medium
fe80::/64 dev eth0.2 proto kernel metric 256 pref medium
fe80::/64 dev br-lan proto kernel metric 256 pref medium
fe80::/64 dev wlan0 proto kernel metric 256 pref medium
local ::1 dev lo table local proto kernel metric 0 pref medium
anycast fe80:: dev eth0 table local proto kernel metric 0 pref medium
anycast fe80:: dev eth0.2 table local proto kernel metric 0 pref medium
anycast fe80:: dev br-lan table local proto kernel metric 0 pref medium
anycast fe80:: dev wlan0 table local proto kernel metric 0 pref medium
local fe80::fa5e:3cff:fe2e:6baa dev eth0 table local proto kernel metric 0 pref medium
local fe80::fa5e:3cff:fe2e:6baa dev eth0.2 table local proto kernel metric 0 pref medium
local fe80::fa5e:3cff:fe2e:6baa dev br-lan table local proto kernel metric 0 pref medium
local fe80::fa5e:3cff:fe2e:6baa dev wlan0 table local proto kernel metric 0 pref medium
ff00::/8 dev eth0 table local proto kernel metric 256 pref medium
ff00::/8 dev eth0.2 table local proto kernel metric 256 pref medium
ff00::/8 dev br-lan table local proto kernel metric 256 pref medium
ff00::/8 dev wlan0 table local proto kernel metric 256 pref medium

`ip address`

1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 1000
    link/ether f8:5e:3c:2e:6b:aa brd ff:ff:ff:ff:ff:ff
    inet6 fe80::fa5e:3cff:fe2e:6baa/64 scope link
       valid_lft forever preferred_lft forever
3: ifb0: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN group default qlen 32
    link/ether 42:e6:56:ac:17:6a brd ff:ff:ff:ff:ff:ff
4: ifb1: <BROADCAST,NOARP> mtu 1500 qdisc noop state DOWN group default qlen 32
    link/ether 6e:67:54:49:e1:a7 brd ff:ff:ff:ff:ff:ff
11: br-lan: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether f8:5e:3c:2e:6b:aa brd ff:ff:ff:ff:ff:ff
    inet 192.168.18.1/24 brd 192.168.18.255 scope global br-lan
       valid_lft forever preferred_lft forever
    inet6 fe80::fa5e:3cff:fe2e:6baa/64 scope link
       valid_lft forever preferred_lft forever
12: eth0.1@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master br-lan state UP group default qlen 1000
    link/ether f8:5e:3c:2e:6b:aa brd ff:ff:ff:ff:ff:ff
13: eth0.2@eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether f8:5e:3c:2e:6b:aa brd ff:ff:ff:ff:ff:ff
    inet 192.168.100.115/24 brd 192.168.100.255 scope global eth0.2
       valid_lft forever preferred_lft forever
    inet6 fe80::fa5e:3cff:fe2e:6baa/64 scope link
       valid_lft forever preferred_lft forever
14: wlan0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq master br-lan state UP group default qlen 1000
    link/ether f8:5e:3c:2e:6b:aa brd ff:ff:ff:ff:ff:ff
    inet6 fe80::fa5e:3cff:fe2e:6baa/64 scope link
       valid_lft forever preferred_lft forever
15: 3g-sim: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UNKNOWN group default qlen 3
    link/ppp
    inet 10.128.114.102 peer 10.64.64.64/32 scope global 3g-sim
       valid_lft forever preferred_lft forever

**Comment by tobiasbrunner** [comment]
> The problem is when WAN1 gets down the IPSec VPN doesn't shift to WAN2 because I think it follows the kernel routing table and tries to establish VPN through WAN1.

It's not really down then, is it? And yes, if the original interface/address/route is still there, strongSwan won't change anything (if the metrics changed it might be possible to switch with `charon.prefer_best_path`).

**Reply by patrakov** [reply]
mwan3 does not change metrics. It inserts IP rules, so the default routing table is irrelevant.
